### ISSUE TITLE
Add peak memory tracking to all benchmarks

### ISSUE THREAD

**Issue by vyasr** [rapidsai/cudf#10528]
#7770 added support for peak memory usage to cuIO benchmarks using rmm's `statistics_resource_adapter`. It would be nice to be able to expand that to all of our benchmarks so that we could more easily detect regressions in memory usage. This would be particularly useful for the Dask cuDF team, which is always looking to identify bottlenecks from memory usage. [There was already discussion](https://github.com/rapidsai/cudf/pull/7770#issuecomment-874473349) of doing this in #7770, so we should investigate following up now.

**Comment by vyasr** [comment]
CC @galipremsagar who was interested in this data.

@devavret it looks like you made this very easy with the `memory_stats_logger`? @harrism would you still be in favor of this?

**Comment by harrism** [comment]
> @harrism would you still be in favor of this?

I don't have a reason not to support this. However it may not benefit the utility of all benchmarks. Does it impact CI benchmark throughput?

**Comment by jrhemstad** [comment]
I am reluctant to add this new functionality based on Google Benchmark when we are trying to phase that out. 

I would support adding this as a feature to benchmarks ported to NVBench.

**Comment by PointKernel** [comment]
> It looks like it's no more complex than incorporating the resource adapter into GBench benchmarks?

Right. It's effortless with the help of [memory_stats_logger](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/benchmarks/fixture/benchmark_fixture.hpp#L103-L122). Only two lines to be added:
```cpp
auto mem_stats_logger = cudf::memory_stats_logger(); // init stats logger
state.exec([&](nvbench::launch& launch) {
  target_kernel();
});
state.add_element_count(mem_stats_logger.peak_memory_usage(), "Peak Memory"); // report peak memory usage to nvbench

```

**Comment by vyasr** [comment]
So then maybe making this change is independent of whether a benchmark has been switch from GBench to NVBench? It seems like we could add this now, then when a project switches from GBench to NVBench the only required change are to switch `stat.add_element_count(` for `state.counters["peak_memory_usage"]`.

**Comment by jrhemstad** [comment]
> So then maybe making this change is independent of whether a benchmark has been switch from GBench to NVBench?

There's also the parsing and reporting side that will be different.

I don't want to keep building things on top of GBench when I'm actively trying to get people to switch to NVbench.

**Comment by vyasr** [comment]
@GregoryKimball do you think this is worth doing systematically for all benchmarks, or something we can add on a case by case basis as the need arises? If the latter, I would vote that we close this issue and just add this when we are benchmarking specific functionality.

**Comment by lhlRahman** [comment]
Hi @GregoryKimball, picking this up as discussed. Plan is one PR per benchmark directory, starting with sort/. Could you assign me?

**Comment by vyasr** [comment]
@lhlRahman assigned!
